docs/man/memcp.1
.TH "MEMCP" "1" "September 16, 2012" "1.0.10" "libmemcached"
.SH NAME
memcp \- libmemcached Documentation
.
.nr rst2man-indent-level 0
.
.de1 rstReportMargin
\\$1 \\n[an-margin]
level \\n[rst2man-indent-level]
level margin: \\n[rst2man-indent\\n[rst2man-indent-level]]
-
\\n[rst2man-indent0]
\\n[rst2man-indent1]
\\n[rst2man-indent2]
..
.de1 INDENT
.\" .rstReportMargin pre:
. RS \\$1
. nr rst2man-indent\\n[rst2man-indent-level] \\n[an-margin]
. nr rst2man-indent-level +1
.\" .rstReportMargin post:
..
.de UNINDENT
. RE
.\" indent \\n[an-margin]
.\" old: \\n[rst2man-indent\\n[rst2man-indent-level]]
.nr rst2man-indent-level -1
.\" new: \\n[rst2man-indent\\n[rst2man-indent-level]]
.in \\n[rst2man-indent\\n[rst2man-indent-level]]u
..
.\" Man page generated from reStructeredText.
.
.sp
Copies files to a collection of memcached servers
.SH SYNOPSIS
.sp
memcp [options] [file] [server]
.SH DESCRIPTION
.sp
\fBmemcp\fP copies one or more files into memcached(1) servers.
It is similar to the standard UNIX cp(1) command.
.sp
The key names will be the names of the files,
without any directory path.
.SH OPTIONS
.sp
You can specify servers via the option:
.INDENT 0.0
.TP
.B \-\-servers
.UNINDENT
.sp
or via the environment variable:
.INDENT 0.0
.TP
.B \(gaMEMCACHED_SERVERS\(ga
.UNINDENT
.sp
If you do not specify either these, the final value in the command line list is the name of a server(s).
.sp
For a full list of operations run the tool with the option:
.INDENT 0.0
.TP
.B \-\-help
.UNINDENT
.SH HOME
.sp
To find out more information please check:
\fI\%http://libmemcached.org/\fP
.SH AUTHOR
.sp
Brian Aker, <\fI\%brian@tangent.org\fP>
.sp
Mark Atwood, <\fI\%mark@fallenpegasus.com\fP>
.SH SEE ALSO
.sp
\fImemcached(1)\fP \fIlibmemcached(3)\fP
.SH AUTHOR
Brian Aker
.SH COPYRIGHT
2011-2012, Brian Aker DataDifferential, http://datadifferential.com/
.\" Generated by docutils manpage writer.
.\" 
.


ChangeLog
1.0.11 Sun Sep 16 20:32:13 EDT 2012
* Removed custom version of memcached.
* Updated hardening rules.
* Fixed a case where the return error from a socket connection differred from that of a TCP/IP socket.

1.0.10 Sun Jul 29 21:50:15 PDT 2012
* --disable-assert has been removed from configure, and --enable-assert has been added in its place.
* Compiling fixes for Clang on OSX Mountain Lion.

1.0.9 Wed Jul  4 22:46:19 EDT 2012
* Faster close on socket.
* Instance allocation is now seperated from server interface. This should allow for a better preservation of ABI compliance from now on.
* Fix close on exec bug.
* Numerous other bug fixes.

1.0.8 Tue May 22 15:06:04 EDT 2012
* Added support for setting options via ENV variable LIBMEMCACHED
* Fix corner case on last used result.

1.0.7 Sat Apr 28 00:48:29 PDT 2012
* Add API call for exist calls.
* Update all license files to be BSD.

1.0.6 Sat Apr  7 18:26:49 PDT 2012
* Fixes for gcc 4.7, lp:961812
* Fix for restart issue that happens under testing.
* Fix for lp:962815.
* Support for transparent AES encryption.

1.0.5 Tue Mar 13 22:56:47 PDT 2012
* Fixes for OSX.
* Version is now parsed directly in the parser, which makes buffered operations now work with it..
* memstat has been extended so that it can be used to find the version of the server.
* Update documentation.
* Fixes for compile issues on Debian and Ubuntu

1.0.4 Thu Jan 26 22:33:54 PST 2012
* Fix for memcached_dump().
* Additional testing for memcached_stat_execute().

1.0.3 Sun Jan  8 18:11:36 PST 2012
* Increased size of sort buffer used during Ketama.
* Added support for new behavior to handle dead servers.
* Overall haul of UDP IO.
* Fixed C compile issue with memcached_exist()
* Numerous bug fixes.
* Clang support for OSX.
* All commands now using vector send support.

1.0.2 Mon Oct 24 08:00:48 PDT 2011
* Dropped libmemcached/memcached_util.h (undocumented header file)
* Added memcached_touch() and memcached_touch_by_key()
* UDP support restructured to toggle on a complete memcached_st structure.

0.53 Mon Sep 26 20:50:33 PDT 2011
  * Fix for FreeBSD/OpenBSD and -lm
  * Added memcached_exist()
  * Fix for memory when using config test.
  * CLI gained --quiet

0.52 Sun Sep 11 22:16:08 PDT 2011
  * Build fixes for Ubuntu/Suse.
  * Fixes for OSX Lion.
  * Bug fix for looping back through dns lookups under certain failures.
  * Fixes related to dead server failures.

0.51 Thu Jul 21 11:08:58 PDT 2011
  * memcached_callback_set() now takes its data argument as const
  * Update to tests.
  * Fix in parser for port number.

0.50 Mon Jun 20 10:36:57 PDT 2011
 * Updates to C++ interface 
 * Custom free allocators need to now check for value before calling free.
 * memcached_fetch_result() now uses the internal result when available (about 25 to 50% faster).
 * Fix for stats structure.
 * Updates to documentation.
 * memcached_fetch_result() now uses the internal result when available (about 25 to 50% faster).

0.49 Thu Apr 14 08:43:37 PDT 2011
  * Fix calls to auto methods so that if value is not passed in nothing bad happens.
  * New parser calls for generating memcached_st objects.
  * New error system.
  * New flow control for messages means faster get/set calls.
  * Added new documentation system.
  * A behavior change has been now made that if you specify a weight for any server, we enable the weight flag and do weight balancing.  
  * Added MEMCACHED_BEHAVIOR_REMOVE_FAILED_SERVERS to simplify the setting of AUTO REJECT for servers.

0.48 Tue Mar 15 23:05:18 PDT 2011
  * Fix memory leak in server parse.
  * Move test framework out to be its own library (easier to work with Gearman).


0.47 Wed Feb 23 18:46:48 PST 2011
  * Additional fixes for OpenBSD.
  * Bug fix 677609, 456080.
  * SIGPIPE fix for Linux send(). 
  * memcapable can now test ascii or binary based on flags.
  * Additional build fixes for SASL.


0.46 Mon Feb 14 10:28:01 PST 2011
  * Fixes a number of corner case bugs.
  * Fixes related to OpenBSD.
  * Better testing for protocol version.
  * Removes special case infinite wait on blocking setup.

0.45 Tue Feb  8 16:02:06 PST 2011
  * Add support for systemtap

0.44 Wed Sep 22 21:57:57 PDT 2010
  * Windows bug fixes.
  * Hudson port support in test harness.
  * Improved portability of test hanrness.
  * SASL fixes.

0.43 Wed Jul 28 16:29:47 PDT 2010
  * Added --args to memstat so that a greater range of values can be returned.
  * Prelimanary support for Windows.
  * memcached_stat_execute() merged. 

0.42 Tue Jul  6 12:29:50 PDT 2010
  * Mistake in libtool caused issue with library version

0.41 Wed Jun 30 13:15:19 PDT 2010
  * Added --file for memcat.
  * Added limemcached_ping() to libmemcached_util
  * Bugfix for some cases where connect would have issues with timeout.
  * Wrong value for errno given as error on an IO failure inside of poll.
  * Bug fix for issue where multiple interfaces with bad DNS were not being caught.

0.40 Thu Apr 22 19:01:25 PDT 2010
  * Placed retry logic in for busted resolvers
  * Add an ignore for SIGPIPE to solve OSX issues.
  * A couple of fixed for memcached_light server.
  * Updated to debug mode to track io_wait

0.39 Tue Apr  6 12:35:13 PDT 2010
  * Add support for prefix keys to binary protocol.
  * Remove the undocumented call memcached_server_remove().
  * The undocumented call memcached_server_by_key() now returns const.
  * memcached_server_error_reset() has been deprecated.
  * memcached_server_list() has been deprecated. Use memcached_server_cursor() to walk the servers found in a memcached_st() structure.
  * memcached_verbosity() can now be run concurrently with other operations.
  * SASL support.
  * Fixes memory leak found in EJECT HOSTS.

0.38 Wed Feb 10 09:40:33 PST 2010
  * C++ interface for libhashkit.
  * Modified memcached_set_memory_allocators() so that it requires a context pointer.
  * memcached_clone() now runs 5 times faster.
  * Functions used for callbacks are now given const memcached_st.

clients/memcp.cc
/* LibMemcached
 * Copyright (C) 2006-2009 Brian Aker
 * All rights reserved.
 *
 * Use and distribution licensed under the BSD license.  See
 * the COPYING file in the parent directory for full text.
 *
 * Summary:
 *
 */

#include "config.h"

#include <cerrno>
#include <climits>
#include <cstdio>
#include <cstdlib>
#include <cstdlib>
#include <cstring>
#include <fcntl.h>
#include <getopt.h>
#include <iostream>
#ifdef HAVE_STRINGS_H
#include <strings.h>
#endif
#include <sys/stat.h>
#include <sys/types.h>
#include <sys/types.h>
#include <sys/types.h>
#include <unistd.h>


#include <libmemcached/memcached.h>

#include "client_options.h"
#include "utilities.h"

#define PROGRAM_NAME "memcp"
#define PROGRAM_DESCRIPTION "Copy a set of files to a memcached cluster."

/* Prototypes */
static void options_parse(int argc, char *argv[]);

static bool opt_binary= false;
static bool opt_udp= false;
static bool opt_buffer= false;
static int opt_verbose= 0;
static char *opt_servers= NULL;
static char *opt_hash= NULL;
static int opt_method= OPT_SET;
static uint32_t opt_flags= 0;
static time_t opt_expires= 0;
static char *opt_username;
static char *opt_passwd;

static long strtol_wrapper(const char *nptr, int base, bool *error)
{
  long val;
  char *endptr;

  errno= 0;    /* To distinguish success/failure after call */
  val= strtol(nptr, &endptr, base);

  /* Check for various possible errors */

  if ((errno == ERANGE and (val == LONG_MAX or val == LONG_MIN))
      or (errno != 0 && val == 0))
  {
    *error= true;
    return EXIT_SUCCESS;
  }

  if (endptr == nptr)
  {
    *error= true;
    return EXIT_SUCCESS;
  }

  *error= false;
  return val;
}

int main(int argc, char *argv[])
{

  options_parse(argc, argv);
  initialize_sockets();

  memcached_st *memc= memcached_create(NULL);

  if (opt_udp)
  {
    if (opt_verbose)
    {
      std::cout << "Enabling UDP" << std::endl;
    }

    if (memcached_failed(memcached_behavior_set(memc, MEMCACHED_BEHAVIOR_USE_UDP, opt_udp)))
    {
      memcached_free(memc);
      std::cerr << "Could not enable UDP protocol." << std::endl;
      return EXIT_FAILURE;
    }
  }

  if (opt_buffer)
  {
    if (opt_verbose)
    {
      std::cout << "Enabling MEMCACHED_BEHAVIOR_BUFFER_REQUESTS" << std::endl;
    }

    if (memcached_failed(memcached_behavior_set(memc, MEMCACHED_BEHAVIOR_BUFFER_REQUESTS, opt_buffer)))
    {
      memcached_free(memc);
      std::cerr << "Could not enable MEMCACHED_BEHAVIOR_BUFFER_REQUESTS." << std::endl;
      return EXIT_FAILURE;
    }
  }

  process_hash_option(memc, opt_hash);

  if (opt_servers == NULL)
  {
    char *temp;

    if ((temp= getenv("MEMCACHED_SERVERS")))
    {
      opt_servers= strdup(temp);
    }
    else
    {
      std::cerr << "No Servers provided" << std::endl;
      exit(EXIT_FAILURE);
    }
  }

  memcached_server_st *servers;
  if (opt_servers)
  {
    servers= memcached_servers_parse(opt_servers);
  }
  else
  {
    servers= memcached_servers_parse(argv[--argc]);
  }

  memcached_server_push(memc, servers);
  memcached_server_list_free(servers);
  memcached_behavior_set(memc, MEMCACHED_BEHAVIOR_BINARY_PROTOCOL, opt_binary);
  if (opt_username and LIBMEMCACHED_WITH_SASL_SUPPORT == 0)
  {
    memcached_free(memc);
    std::cerr << "--username was supplied, but binary was not built with SASL support." << std::endl;
    return EXIT_FAILURE;
  }

  if (opt_username)
  {
    memcached_return_t ret;
    if (memcached_failed(ret= memcached_set_sasl_auth_data(memc, opt_username, opt_passwd)))
    {
      std::cerr << memcached_last_error_message(memc) << std::endl;
      memcached_free(memc);
      return EXIT_FAILURE;
    }
  }

  int exit_code= EXIT_SUCCESS;
  while (optind < argc)
  {
    int fd= open(argv[optind], O_RDONLY);
    if (fd < 0)
    {
      if (opt_verbose)
      {
        std::cerr << "memcp " << argv[optind] << " " << strerror(errno) << std::endl;
        optind++;
      }
      exit_code= EXIT_FAILURE;
      continue;
    }

    struct stat sbuf;
    (void)fstat(fd, &sbuf);

    char *ptr= rindex(argv[optind], '/');
    if (ptr)
    {
      ptr++;
    }
    else
    {
      ptr= argv[optind];
    }

    if (opt_verbose)
    {
      static const char *opstr[] = { "set", "add", "replace" };
      printf("op: %s\nsource file: %s\nlength: %lu\n"
	     "key: %s\nflags: %x\nexpires: %lu\n",
	     opstr[opt_method - OPT_SET], argv[optind], (unsigned long)sbuf.st_size,
	     ptr, opt_flags, (unsigned long)opt_expires);
    }

    char *file_buffer_ptr;
    if ((file_buffer_ptr= (char *)malloc(sizeof(char) * (size_t)sbuf.st_size)) == NULL)
    {
      std::cerr << "Error allocating file buffer(" << strerror(errno) << ")" << std::endl;
      close(fd);
      exit(EXIT_FAILURE);
    }

    ssize_t read_length;
    if ((read_length= ::read(fd, file_buffer_ptr, (size_t)sbuf.st_size)) == -1)
    {
      std::cerr << "Error while reading file " << file_buffer_ptr << " (" << strerror(errno) << ")" << std::endl;
      close(fd);
      exit(EXIT_FAILURE);
    }

    if (read_length != sbuf.st_size)
    {
      std::cerr << "Failure while reading file. Read length was not equal to stat() length" << std::endl;
      close(fd);
      exit(EXIT_FAILURE);
    }

    memcached_return_t rc;
    if (opt_method == OPT_ADD)
    {
      rc= memcached_add(memc, ptr, strlen(ptr),
                        file_buffer_ptr, (size_t)sbuf.st_size,
			opt_expires, opt_flags);
    }
    else if (opt_method == OPT_REPLACE)
    {
      rc= memcached_replace(memc, ptr, strlen(ptr),
			    file_buffer_ptr, (size_t)sbuf.st_size,
			    opt_expires, opt_flags);
    }
    else
    {
      rc= memcached_set(memc, ptr, strlen(ptr),
                        file_buffer_ptr, (size_t)sbuf.st_size,
                        opt_expires, opt_flags);
    }

    if (memcached_failed(rc))
    {
      std::cerr << "Error occrrured during memcached_set(): " << memcached_last_error_message(memc) << std::endl;
      exit_code= EXIT_FAILURE;
    }

    ::free(file_buffer_ptr);
    ::close(fd);
    optind++;
  }

  if (opt_verbose)
  {
    std::cout << "Calling memcached_free()" << std::endl;
  }

  memcached_free(memc);

  if (opt_servers)
  {
    free(opt_servers);
  }

  if (opt_hash)
  {
    free(opt_hash);
  }

  return exit_code;
}

static void options_parse(int argc, char *argv[])
{
  memcached_programs_help_st help_options[]=
  {
    {0},
  };

  static struct option long_options[]=
    {
      {(OPTIONSTRING)"version", no_argument, NULL, OPT_VERSION},
      {(OPTIONSTRING)"help", no_argument, NULL, OPT_HELP},
      {(OPTIONSTRING)"quiet", no_argument, NULL, OPT_QUIET},
      {(OPTIONSTRING)"udp", no_argument, NULL, OPT_UDP},
      {(OPTIONSTRING)"buffer", no_argument, NULL, OPT_BUFFER},
      {(OPTIONSTRING)"verbose", no_argument, &opt_verbose, OPT_VERBOSE},
      {(OPTIONSTRING)"debug", no_argument, &opt_verbose, OPT_DEBUG},
      {(OPTIONSTRING)"servers", required_argument, NULL, OPT_SERVERS},
      {(OPTIONSTRING)"flag", required_argument, NULL, OPT_FLAG},
      {(OPTIONSTRING)"expire", required_argument, NULL, OPT_EXPIRE},
      {(OPTIONSTRING)"set",  no_argument, NULL, OPT_SET},
      {(OPTIONSTRING)"add",  no_argument, NULL, OPT_ADD},
      {(OPTIONSTRING)"replace",  no_argument, NULL, OPT_REPLACE},
      {(OPTIONSTRING)"hash", required_argument, NULL, OPT_HASH},
      {(OPTIONSTRING)"binary", no_argument, NULL, OPT_BINARY},
      {(OPTIONSTRING)"username", required_argument, NULL, OPT_USERNAME},
      {(OPTIONSTRING)"password", required_argument, NULL, OPT_PASSWD},
      {0, 0, 0, 0},
    };

  bool opt_version= false;
  bool opt_help= false;
  int option_index= 0;

  while (1)
  {
    int option_rv= getopt_long(argc, argv, "Vhvds:", long_options, &option_index);

    if (option_rv == -1)
      break;

    switch (option_rv)
    {
    case 0:
      break;

    case OPT_BINARY:
      opt_binary= true;
      break;

    case OPT_VERBOSE: /* --verbose or -v */
      opt_verbose= OPT_VERBOSE;
      break;

    case OPT_DEBUG: /* --debug or -d */
      opt_verbose= OPT_DEBUG;
      break;

    case OPT_VERSION: /* --version or -V */
      opt_version= true;
      break;

    case OPT_HELP: /* --help or -h */
      opt_help= true;
      break;

    case OPT_SERVERS: /* --servers or -s */
      opt_servers= strdup(optarg);
      break;

    case OPT_FLAG: /* --flag */
      {
        bool strtol_error;
        opt_flags= (uint32_t)strtol_wrapper(optarg, 16, &strtol_error);
        if (strtol_error == true)
        {
          fprintf(stderr, "Bad value passed via --flag\n");
          exit(1);
        }
      }
      break;

    case OPT_EXPIRE: /* --expire */
      {
        bool strtol_error;
        opt_expires= (time_t)strtol_wrapper(optarg, 16, &strtol_error);
        if (strtol_error == true)
        {
          fprintf(stderr, "Bad value passed via --flag\n");
          exit(1);
        }
      }
      break;

    case OPT_SET:
      opt_method= OPT_SET;
      break;

    case OPT_REPLACE:
      opt_method= OPT_REPLACE;
      break;

    case OPT_ADD:
      opt_method= OPT_ADD;
      break;

    case OPT_HASH:
      opt_hash= strdup(optarg);
      break;

    case OPT_USERNAME:
      opt_username= optarg;
      break;

    case OPT_PASSWD:
      opt_passwd= optarg;
      break;

    case OPT_QUIET:
      close_stdio();
      break;

    case OPT_UDP:
      opt_udp= true;
      break;

    case OPT_BUFFER:
      opt_buffer= true;
      break;

    case '?':
      /* getopt_long already printed an error message. */
      exit(1);
    default:
      abort();
    }
  }

  if (opt_version)
  {
    version_command(PROGRAM_NAME);
    exit(EXIT_SUCCESS);
  }

  if (opt_help)
  {
    help_command(PROGRAM_NAME, PROGRAM_DESCRIPTION, long_options, help_options);
    exit(EXIT_SUCCESS);
  }
}
